### PR TITLE
chore: separating DirectX code from encoder code

### PR DESCRIPTION
**Pull request by sindharta**

### CODE REVIEW

File: Plugin~/WebRTCPlugin/GraphicsDevice/IGraphicsDevice.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+class IGraphicsDevice {
+public:
+
+    IGraphicsDevice();
+    virtual ~IGraphicsDevice() = 0;
+
+    virtual void ShutdownV() = 0;
+    virtual ITexture2D* CreateEncoderInputTextureV(uint32_t width, uint32_t height) = 0;
+    virtual ITexture2D* CreateEncoderInputTextureV(uint32_t width, uint32_t height, void* nativeTexturePtr) = 0;
+    virtual void* GetNativeDevicePtrV() = 0;
```
Comment (karasusan) [inline Plugin~/WebRTCPlugin/GraphicsDevice/IGraphicsDevice.h]: It seems better this method is declared as `const`

File: Plugin~/WebRTCPlugin/GraphicsDevice/ITexture2D.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace WebRTC {
+
+struct ITexture2D {
+public:
+    //[TODO-Sin: 2019-19-11] ITexture2D should not be created directly, but should be called using
+    //GraphicsDevice->CreateEncoderInputTexture
+
+    ITexture2D(uint32_t w, uint32_t h) : m_width (w), m_height(h) {}
+    bool IsSize(uint32_t w, uint32_t h) { return m_width == w && m_height == h; }
```
Comment (karasusan) [inline Plugin~/WebRTCPlugin/GraphicsDevice/ITexture2D.h]: It seems better this method is declared as `const`

File: Plugin~/WebRTCPlugin/GraphicsDevice/D3D12/D3D12Texture2D.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+public:
+    ID3D12Resource* m_texture;
+
+    D3D12Texture2D(uint32_t w, uint32_t h,ID3D12Resource* tex);
+
+    virtual ~D3D12Texture2D() {
+        SAFE_RELEASE(m_texture);
+    }
+
+    inline virtual void* GetNativeTexturePtrV();
```
[the remark this answers is not in the document]
Comment (sindharta) [reply Plugin~/WebRTCPlugin/GraphicsDevice/D3D12/D3D12Texture2D.h]: This is required for copying texture before encoding.
See `UnityVideoCapturer::EncodeVideoData()`

File: Plugin~/WebRTCPlugin/GraphicsDevice/IGraphicsDevice.h (hunk 1 above)
Comment (sindharta) [reply Plugin~/WebRTCPlugin/GraphicsDevice/IGraphicsDevice.h]: Yeah, I thought the same way, but `NV_ENC_OPEN_ENCODE_SESSION_EX_PARAMS.device` doesn't accept const, so it won't work

File: Plugin~/WebRTCPlugin/GraphicsDevice/ITexture2D.h (hunk 2 above)
Comment (sindharta) [reply Plugin~/WebRTCPlugin/GraphicsDevice/ITexture2D.h]: ok

File: Plugin~/WebRTCPlugin/GraphicsDevice/IGraphicsDevice.h (hunk 1 above)
Comment (karasusan) [reply Plugin~/WebRTCPlugin/GraphicsDevice/IGraphicsDevice.h]: Got it

**Pull request merged**